### PR TITLE
Added recording of exceptions in Pyramid

### LINKED ISSUE #2619: Exceptions are not recorded in traces from instrumented Pyramid apps
### What problem do you want to solve?

Traces resulting from Pyramid instrumentation do not record server exceptions that have resulted in errored requests. Here is the relevant output part by calling an instrumented Pyramid app.

```
    "status": {
        "status_code": "ERROR"
    },
    "attributes": {
        "http.method": "GET",
        ...
        "http.status_code": 500
    },
    "events": []
```

In case of another WSGI, e.g. Flask, errors are bubbled up in status with the traceback as a span event looking like this

```
    "status": {
        "status_code": "ERROR",
        "description": "ValueError: bad value"
    },
    "attributes": {
        "http.method": "GET",
        ...
        "http.status_code": 500
    },
    "events": [
        {
            "name": "exception",
            "timestamp": "2024-06-19T10:44:00.468880Z",
            "attributes": {
                "exception.type": "ValueError",
                "exception.message": "bad value",
                "exception.stacktrace": "Traceback ...",
                "exception.escaped": "False"
            }
        }
    ]
```

Such traces are much more valuable in troubleshooting.

### Describe the solution you'd like

Implement recording of exceptions as status attributes and span events in Pyramid instrumentation, similar to how it is done for other WSGI apps.

### Describe alternatives you've considered

_No response_

### Additional Context

_No response_

### Would you like to implement a fix?

Yes

### PR DESCRIPTION
**Pull request by rbagd**
# Description

This PR adapts the instrumentation of Pyramid applications so that traces would contain information on uncaught exceptions or errors while serving a request. This is similar to what is done in the generic `OpenTelemetryMiddleware` for WSGI apps, which is not used in Pyramid instrumentation.

Fixes #2619

## Type of change

- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Unit tests added to ensure information on exceptions is present in generated traces. Only HTTP 5xx-like responses should generate such exceptions.

# Does This PR Require a Core Repo Change?

- [X] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [X] Followed the style guidelines of this project
- [X] Changelogs have been updated
- [X] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-pyramid/tests/test_programmatic.py [hunk 1]
```diff
@@ -166,6 +168,18 @@ def test_internal_error(self):
         self.assertEqual(span_list[0].name, "/hello/{helloid}")
         self.assertEqual(span_list[0].kind, trace.SpanKind.SERVER)
         self.assertEqual(span_list[0].attributes, expected_attrs)
+        self.assertEqual(
+            span_list[0].status.status_code, trace.StatusCode.ERROR
+        )
+        self.assertLess(
```
Comment (xrmx) [inline instrumentation/opentelemetry-instrumentation-pyramid/tests/test_programmatic.py:174]: If you are aim is to assert that a string is in another maybe `assertIn`?

File: instrumentation/opentelemetry-instrumentation-pyramid/tests/test_programmatic.py [hunk 2]
```diff
@@ -184,6 +198,20 @@ def test_internal_exception(self):
...
+            span_list[0].status.status_code, trace.StatusCode.ERROR
+        )
+        self.assertEqual(span_list[0].status.description, "error message")
+
+        expected_error_event_attrs = {
+            exception_attributes.EXCEPTION_TYPE: "NotImplementedError",
+            exception_attributes.EXCEPTION_MESSAGE: "error message",
+        }
+        self.assertEqual(span_list[0].events[0].name, "exception")
+        self.assertLess(
```
Comment (xrmx) [inline instrumentation/opentelemetry-instrumentation-pyramid/tests/test_programmatic.py:211]: That is to assert that `expected_error_event_attrs` is a subset of `dict(span_list[0].events[0].attributes)`? Could we make it more explicit?

Comment (xrmx) [reply instrumentation/opentelemetry-instrumentation-pyramid/tests/test_programmatic.py:211]: Maybe it's just me not used to use `<` to compare dictionary items :sweat_smile:

Comment (rbagd) [reply instrumentation/opentelemetry-instrumentation-pyramid/tests/test_programmatic.py:211]: Unfortunately, `assertDictContainsSubset` got deprecated. :cry: Maybe I can add a comment to give a brief explanation.

Comment (xrmx) [reply instrumentation/opentelemetry-instrumentation-pyramid/tests/test_programmatic.py:212]: Great, thanks

**Review by xrmx — APPROVED** [review]
Thanks!

**Review by lzchen — APPROVED** [review]

**Pull request merged**